### PR TITLE
Port prefix/cd context managers for run/sudo 

### PR DESCRIPTION
**Pull request by rpkilby**

### CODE REVIEW

File: invoke/context.py [hunk 1]
```diff
@@ -192,6 +202,134 @@ def sudo(self, command, **kwargs):
...
+        """
+        Prefix all nested `run`/`sudo` commands with given command plus ``&&``.
+
+        Most of the time, you'll want to be using this alongside a shell script
+        which alters shell state, such as ones which export or alter shell
+        environment variables.
+
+        For example, one of the most common uses of this tool is with the
+        ``workon`` command from `virtualenvwrapper
+        <http://www.doughellmann.com/projects/virtualenvwrapper/>`_::
```
Comment (bitprophet) [inline invoke/context.py]: This is now https://virtualenvwrapper.readthedocs.io so we should probs update it!

File: invoke/context.py [hunk 2]
```diff
@@ -192,6 +202,134 @@ def sudo(self, command, **kwargs):
...
+        """
+        if not self.command_cwds:
+            return ''
+
+        # get the index for the subset of paths starting with the last / or ~
+        for i, path in reversed(list(enumerate(self.command_cwds))):
+            if path.startswith('~') or path.startswith('/'):
+                break
+
+        paths = [path.replace(' ', '\ ') for path in self.command_cwds[i:]]
```
Comment (bitprophet) [inline invoke/context.py]: Forget offhand but we may want to see if there's a stronger "escape this path" function somewhere we can reuse - think escaping tildes or slashes in filenames and dumb shit like that.

**Comment by rpkilby** [comment]
All passing on CI now.

**Comment by bitprophet** [comment]
a) your avatar is hilarious in a Github context, b) haven't forgotten about this! Still doing post-PyCon triage but I have this open in a "deal with it soon" browser tab... 😒

**Comment by rpkilby** [comment]
No worries. I'm aware of the very many plates you have 😄

**Comment by bitprophet** [comment]
Also, realized there were two other open tickets about this, that I can now close. Woo! Thanks yet again.

**Pull request merged**